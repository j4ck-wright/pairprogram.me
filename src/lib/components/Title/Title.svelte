<script lang="ts">
	import { roomTitleStore, setTitle, watchTitle } from '$lib/firebase';

	const MAX_CHARACTERS = 32;
	let previousTitle = $roomTitleStore as string;
	let editHint = false;
	let active = false;
	let error = false;
	let errorMessage: string;

	function toggleActive() {
		active = !active;
	}

	function validTitle() {
		if (!$roomTitleStore && !previousTitle) return true;
		return $roomTitleStore.length > 0 && $roomTitleStore.length < MAX_CHARACTERS;
	}

	watchTitle();
</script>

<!-- svelte-ignore a11y-no-noninteractive-element-interactions -->
<!-- svelte-ignore a11y-click-events-have-key-events -->
<h3
	class="text-3xl mb-2 font-bold cursor-pointer w-max tooltip-error tooltip-left {error &&
		'tooltip tooltip-open'}"
	data-tip={errorMessage}
	on:click={toggleActive}
	on:mouseenter={() => {
		if (!active) {
			editHint = !editHint;
		}
	}}
	on:mouseleave={() => {
		if (!active) {
			editHint = false;
		}
	}}
>
	{#if active}
		<!-- svelte-ignore a11y-autofocus -->
		<input
			type="text"
			autofocus
			class="pl-2 outline-primary bg-transparent"
			size="12"
			bind:value={$roomTitleStore}
			on:keydown={(e) => {
				if (e.key === 'Enter' && validTitle()) {
					active = false;
					error = false;
					previousTitle = $roomTitleStore;
				} else if (!validTitle()) {
					error = true;
					errorMessage = `Title must be between 1-${MAX_CHARACTERS} letters`;
				} else {
					error = false;
				}
			}}
			on:focusout={() => {
				active = false;
				editHint = false;
				error = false;
				if (validTitle()) {
					previousTitle = $roomTitleStore;
					setTitle($roomTitleStore);
				} else {
					$roomTitleStore = previousTitle;
				}
			}}
		/>
	{:else}
		{$roomTitleStore}
	{/if}
	<iconify-icon
		role="button"
		icon="mdi:pencil"
		class="text-primary text-2xl bg-transparent text-opacity-75 {!editHint && 'invisible'}"
	/>
</h3>
