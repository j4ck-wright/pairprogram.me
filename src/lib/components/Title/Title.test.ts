import { render, screen } from '@testing-library/svelte';
import Title from './Title.svelte';

describe('Title Test Suite', () => {
	beforeEach(() => {
		render(Title, {
			title: 'Test title'
		});
	});

	it('Correctly displays the title when passed in as a prop', () => {
		expect(screen.getByText('Test title')).toBeTruthy();
	});
});

describe('ParticipantBadge snapshot', () => {
	it('Should match snapshot', () => {
		const participantComponent = render(Title);
		expect(participantComponent).toMatchSnapshot();
	});
});
