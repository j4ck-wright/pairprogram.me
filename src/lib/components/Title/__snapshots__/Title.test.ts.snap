// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ParticipantBadge snapshot > Should match snapshot 1`] = `
{
  "component": Title {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        undefined,
        undefined,
        false,
        false,
        false,
        undefined,
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "title": 0,
      },
      "root": <div>
        <h3
          class="text-3xl mb-2 font-bold cursor-pointer w-max tooltip-error tooltip-left false"
        >
          undefined
           
          <iconify-icon
            class="text-primary text-2xl bg-transparent text-opacity-75 invisible"
            icon="mdi:pencil"
            role="button"
          />
        </h3>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$events_def": undefined,
    "$$prop_def": undefined,
    "$$set": [Function],
    "$$slot_def": undefined,
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <h3
        class="text-3xl mb-2 font-bold cursor-pointer w-max tooltip-error tooltip-left false"
      >
        undefined
         
        <iconify-icon
          class="text-primary text-2xl bg-transparent text-opacity-75 invisible"
          icon="mdi:pencil"
          role="button"
        />
      </h3>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
